src/index.js

```diff
@@ -29,26 +50,63 @@ const cors = (options = {}) => handler => (req, res, ...restArgs) => {
     runHandlerOnOptionsRequest = false
   } = options
 
-  res.setHeader('Access-Control-Allow-Origin', origin)
-  if (allowCredentials) {
-    res.setHeader('Access-Control-Allow-Credentials', 'true')
-  }
-  if (exposeHeaders.length) {
-    res.setHeader('Access-Control-Expose-Headers', exposeHeaders.join(','))
+  const requestOrigin = req.headers.origin
+
+  if (!requestOrigin) {
+    return handler(req, res, ...restArgs)
   }
 
-  const preFlight = req.method === 'OPTIONS'
-  if (preFlight) {
-    res.setHeader('Access-Control-Allow-Methods', allowMethods.join(','))
-    res.setHeader('Access-Control-Allow-Headers', allowHeaders.join(','))
-    res.setHeader('Access-Control-Max-Age', String(maxAge))
+  let originCallback = null
```

Comment:
Could you please describe the `originCallback` concept?

Reply:
yes, that part is a little more complicated because of the possibility that `options.origin` is a function.
so we first check if `options.origin` is a function. if it is a function we use this function as `originCallback`
if `options.origin` is not empty but not a function we use our own function that just calls the callback with the origin as it's second parameter.

after this we have to call `originCallback` to handle the request correctly.

maybe it's clearer when looking at the (still missing) documentation: 
https://github.com/expressjs/cors#configuration-options

in my own words the API for the user looks like this:
`options.origin` accepts the following function: function(requestOrigin, callback)
where `requestOrigin` is the origin from the request and `callback` has the following signature: function(error, allow)

i had different approaches for the function option. that's probably why i declared the function as async before.

Reply:
Ah, I see, thanks for the explanation. 

Since micro is async by default, your original intuition for async was correct. The `expressjs` callback approach doesn't fit this style. 

What do you think about holding off on this feature for now and putting it in its own PR? Then we could have some discussion and design an elegant solution.

Reply:
i think it's a good idea to put this in its own PR. because changing it later could introduce breaking changes. i guess the new options (especially regex) are powerful enough for the most common use-cases. should we start with this when this PR is merged?

btw another idea was to just accept a promise and wait for it to resolve.

should i for now just remove function support and the corresponding tests?

Reply:
Yeah, let’s remove for now.